pkg/elasticetcd/volunteer.go

```diff
@@ -0,0 +1,71 @@
+package elasticetcd
+
+import (
+	"time"
+
+	"github.com/coreos/etcd/clientv3"
+)
+
+// volunteerSelf adds the self to the volunteer list and starts watching for the nomination
+func (ee *ElasticEtcd) volunteerSelf() error {
+	key := volunteerPrefix + ee.conf.Name
+	var val string
+	// Need to set advertisable CURLs here as the initial cluster lists for new
+	// servers will be formed from this, the default CURL is not advertisable.
+	if isDefaultCURL(ee.conf.CURLs) {
+		val = defaultACURLs.String()
+	} else {
+		val = ee.conf.CURLs.String()
+	}
+
+	_, err := ee.cli.Put(ee.cli.Ctx(), key, val, clientv3.WithLease(ee.session.Lease()))
+	if err != nil {
+		ee.log.WithError(err).Error("failed to add self to volunteer list")
+		return err
+	}
+	ee.watchNomination()
+
+	return nil
+}
+
+func (ee *ElasticEtcd) watchNomination() {
+	ee.log.Debug("watching for self nomination")
+
+	key := nomineePrefix + ee.conf.Name
+
+	f := func(_ clientv3.WatchResponse) {
+		ee.handleNomination()
+	}
+	ee.watch(key, f)
+}
+
+func (ee *ElasticEtcd) handleNomination() {
+	ee.lock.Lock()
+	defer ee.lock.Unlock()
+
+	ee.log.Debug("handling nomination")
+
+	// Get the current list of nominees
+	nomineesResp, err := ee.cli.Get(ee.cli.Ctx(), nomineePrefix, clientv3.WithPrefix(), clientv3.WithSort(clientv3.SortByKey, clientv3.SortAscend))
+	if err != nil {
+		ee.log.WithError(err).Error("could not get nominees")
+		return
+	}
+
+	nominees, err := urlsMapFromGetResp(nomineesResp, nomineePrefix)
+	if err != nil {
+		ee.log.WithError(err).Error("could not prepare nominees map")
+	}
+
+	// Check if you are in the nominees, and start/stop you embedded server as
+	// required
+	if _, ok := nominees[ee.conf.Name]; ok {
+		ee.log.Debug("nominated, starting server")
+		// Sleeping to allow leader to add me as a etcd cluster member
+		time.Sleep(2 * time.Second)
+		ee.startServer(nominees.String())
```

Comment:
`ee.startServer`/`ee.stopServer` check internally for the current state of the etcd server before doing their operations. Other than some unnecessary logs, doing this frequently shouldn't be a problem.

Nominations happen on changes to the volunteer list or idealsize. I don't expect these to change very fast. If you are already a server, changes to the volunteer list wouldn't affect you, as any existing online servers would still be volunteered. Your nomination can be removed only when idealsize is lowered. So unless someone keeps bumping an lowering idealsize repeatedly, it shouldn't be a problem as well.

I would like to see some sort of rate limitation for nominations though. I'll a TODO for this as well.